internal/nfs/nodeserver/nodeserver.go

```diff
@@ -0,0 +1,348 @@
+	}
+
+	return nil, status.Errorf(codes.InvalidArgument,
+		"targetpath %q is not a directory or device", targetPath)
+}
+
+// mountNFS mounts nfs volumes.
+func (ns *NodeServer) mountNFS(
+	ctx context.Context,
+	volumeID, clusterID, source, mountPoint string,
+	mountOptions []string,
+	mountPermissions uint64,
+) error {
+	var (
+		stderr, netNamespaceFilePath string
+		err                          error
+	)
+
+	if clusterID != "" {
+		netNamespaceFilePath, err = util.GetNFSNetNamespaceFilePath(
+			util.CsiConfigFile,
+			clusterID)
+		if err != nil {
+			return err
+		}
+	}
+
+	notMnt, err := ns.Mounter.IsLikelyNotMountPoint(mountPoint)
+	if err != nil {
+		if os.IsNotExist(err) {
+			if err = os.MkdirAll(mountPoint, os.FileMode(mountPermissions)); err != nil {
+				return err
+			}
+			notMnt = true
+		} else {
+			return err
+		}
+	}
+	if !notMnt {
+		log.DebugLog(ctx, "nfs: volume is already mounted to %s", mountPoint)
+
+		return nil
+	}
+
+	args := []string{
+		"-t", "nfs",
+		source,
+		mountPoint,
+	}
+
+	if len(mountOptions) > 0 {
+		args = append(append(args, "-o"), mountOptions...)
+	}
+
+	log.DefaultLog("nfs: mounting volumeID(%v) source(%s) targetPath(%s) mountflags(%v)",
+		volumeID, source, mountPoint, mountOptions)
+	if netNamespaceFilePath != "" {
+		_, stderr, err = util.ExecuteCommandWithNSEnter(
+			ctx, netNamespaceFilePath, "mount", args[:]...)
+	} else {
+		_, stderr, err = util.ExecCommand(ctx, "mount", args[:]...)
+	}
+	if err != nil {
+		return fmt.Errorf("nfs: failed to mount %q to %q : %w stderr: %q",
+			source, mountPoint, err, stderr)
+	}
+	if stderr != "" {
+		return fmt.Errorf("nfs: failed to mount %q to %q : stderr %q",
+			source, mountPoint, stderr)
+	}
+
+	return err
+}
+
+// validateNodePublishVoluemRequest validates node publish volume request.
+func validateNodePublishVoluemRequest(req *csi.NodePublishVolumeRequest) error {
+	switch {
+	case req.GetVolumeId() == "":
+		return errors.New("volume ID missing in request")
+	case req.GetVolumeCapability() == nil:
+		return errors.New("volume capability missing in request")
+	case req.GetTargetPath() == "":
+		return errors.New("target path missing in request")
+	case req.GetVolumeContext()[paramServer] == "":
+		return fmt.Errorf("%v missing in request", paramServer)
+	case req.GetVolumeContext()[paramShare] == "":
+		return fmt.Errorf("%v missing in request", paramShare)
+	}
+
+	return nil
+}
+
+// chmodIfPermissionMismatch only performs chmod when permission mismatches.
+func chmodIfPermissionMismatch(
+	ctx context.Context,
+	targetPath string,
+	mode os.FileMode,
+) error {
+	info, err := os.Lstat(targetPath)
+	if err != nil {
+		return err
+	}
+	perm := info.Mode() & os.ModePerm
+	if perm != mode {
+		log.DebugLog(ctx, "performing chmod targetPath(%s, mode:0%o) with permissions(0%o)",
+			targetPath, info.Mode(), mode)
+		if err = os.Chmod(targetPath, mode); err != nil {
+			return err
```

Comment:
just return err from here if we do so we can remove else check

Reply:
done